File: projecteuler/066.py
y = lambda x, D: (x**2 - 1)**0.5 / (D)**0.5

def calculate(D):
    for x in range(2, 1000000):
        z = y(x, D)
        if z == int(z):
            return x
    return 0

squares = (x**2 for x in range(int(1000**.5)+1))
res = max(calculate(d) for d in range(100, 1001) if d not in squares)
print(res)

File: projecteuler/066.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
ll MAX = (1e9);

ll calculate(ll d) {
    ll x = sqrt(d);
    ll y = 1;
    ll left, right;

    while (x < MAX) {
        left = x * x;
        right = 1 + d * y * y;

        if (left == right)
            return x;
        if (left < right)
            ++x;
        else
            ++y;
    }

    return 0;
}

int main() {
    vector<bool> is_square(1007, 0);
    for (int i = 2; i * i <= 1000; ++i)
        is_square[i*i] = 1;

    ll res = 1;
    for (int D = 2; D <= 1000; ++D)
        if (!is_square[D])
            res = max(res, calculate(D)), cout << D << ' ' << calculate(D) << '\n';
    cout << "Solution: " << res << endl;

    return 0;
}
